[manila-csi-plugin]: Implement automatic topology awareness for Manila share provisioning

kayrus opened the pull request:
**What this PR does / why we need it**:

This PR allows to automatically select the corresponding Node AZ without the topology mapping.

**Which issue this PR fixes(if applicable)**:

Fixes #2219

**Release note**:
<!--
1. Release note is required if a significant change is introduced, otherwise please keep this section as is.
2. Release note is in Markdown format and should begin with the binary name unless multiple binaries are affected, e.g. [openstack-cloud-controller-manager] Deprecate Neutron-LBaaS support.
3. Instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Implement automatic topology awareness for Manila share provisioning
```

kayrus commented:
cc @zetaab @gman0 
as for the e2e tests, I'm trying to fix them in #2254

mdbooth commented on pkg/csi/manila/controllerserver.go:
```diff
@@ -135,9 +136,26 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
+	if cs.d.withTopology && accessibleTopologyReq != nil {
+		// All requisite/preferred topologies are considered valid. Nodes from those zones are required to be able to reach the storage.
+		// The operator is responsible for making sure that provided topology keys are valid and present on the nodes of the cluster.
+		accessibleTopology = accessibleTopologyReq.GetPreferred()
```
Reviewers note: this was the previous behaviour.

mdbooth commented on pkg/csi/manila/controllerserver.go:
```diff
@@ -135,9 +136,26 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
+
+		// Check if "availability" parameter is already specified, if not get preferred from Topology
+		// Required, incase share AZ is different from node AZ
+		if shareOpts.AvailabilityZone == "" {
+			shareOpts.AvailabilityZone = util.GetAZFromTopology(topologyKey, accessibleTopologyReq)
```
Reviewers note: If AZ was unset in shareOpts, we now set it based on any topology in the request. This is new behaviour.

mdbooth reviewed (commented):
Can you summarise the cases where existing behaviour would change? i.e. What are all the ways this could break somebody?

kayrus commented:
> Can you summarise the cases where existing behaviour would change? i.e. What are all the ways this could break somebody?

It depends on the defined StorageClass and whether it contains or doesn't contain the `availability` parameter. In short, if the `--with-topology` is set and the `availability` is not set, then AZ is detected automatically according to Nova AZ, following the Cinder CSI logic. This is the only behavior, which will be changed.

From my point of view, if the `availability` parameter isn't maintained, then k8s operator doesn't really care about the availability zones. For rare cases, when you need a strict Manila to Nova AZ mapping, operator should set the `allowedTopology` and this behavior wasn't changed.

kayrus commented:
@mdbooth, is there anything else I can clarify for you?

@jichenjc @zetaab ready for review

I'm not sure whether it's a bug or a feature. Comparing to Cinder CSI behavior, it's a bug. I'd like to make cherry-picks for releases starting from 1.25, since our prod environment needs this fix and currently we have problems migrating to 1.26. Any objections?

mdbooth commented:
I want to try to write in very clear language consumable by a user what the impact is.

I would be nervous about backporting a behavioural change to a stable release. If it's really required, could it be behind a feature gate (e.g. a command line argument) in the backports? We could release note the behaviour change in the next version and drop the feature gate.

kayrus commented:
> If it's really required, could it be behind a feature gate (e.g. a command line argument) in the backports? We could release note the behaviour change in the next version and drop the feature gate.

nice suggestion. I think the easiest way to implement this is in a storage class parameters.

mdbooth commented:
Please excuse me while I ramp up on CSI! I will try to explain my understanding below in very simple terms to both help solidify my own understanding, and so you can hopefully see where I've misunderstood:

This example SC is taken from [the docs](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/manila-csi-plugin/using-manila-csi-plugin.md):
```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: cephfs-gold
provisioner: cephfs.manila.csi.openstack.org
parameters:
  availability: zone-a
  ...
allowedTopologies:
  - matchLabelExpressions:
    - key: topology.manila.csi.openstack.org/zone
      values:
        - nova-1
        - nova-2
```

A volume with this storage class will be created in AZ `zone-a`, and can be attached to any node with labels matching `allowedTopologies`.

Cinder is identical in all relevant respects.

If we set `waitForFirstConsumer`, the create request will additionally have the topology of the node which wants the attachment.

In cinder, if `availability` is unset on the storage class we extract the first matching cinder topology label from the Node and set the availability zone to that. We assume that this is compatible with `allowedTopologies` on the storage class because... otherwise the pod wouldn't have been scheduled there?

In manila, if `availability` was unset we let manila pick.

With this change, if `availability` is unset we used the topology from the target node provided in the create request.

IIUC, this is almost certainly what we want. Who might be broken by this?

Anybody with mislabelled nodes who hadn't noticed because manila always used the default AZ will now see volume creation failures.
Anybody who had become used to the existing behaviour and provisioned sufficient capacity only in the default AZ.

Remediation:
* Explicitly set availability in the storage class
* Explicitly set manila node topology labels to point to the default AZ for all nodes

I'm obviously no expert in this area (working on it), but this sounds reasonable to me. I think the feature gate in the storage class is appropriate. After merging and backporting this fix we should change the default of the feature gate to true and mark it deprecated in master.

kayrus commented:
> In cinder, if availability is unset on the storage class we extract the first matching cinder topology label from the Node and set the availability zone to that. We assume that this is compatible with allowedTopologies on the storage class because... otherwise the pod wouldn't have been scheduled there?

correct. the same logic now implemented in Manila CSI with a new extra `autoTopology` parameter.

> IIUC, this is almost certainly what we want. Who might be broken by this?
> Anybody with mislabelled nodes who hadn't noticed because manila always used the default AZ will now see volume creation failures.

Correct, if `topology-aware` was enabled and storage classes were not maintained enough, there was a chance to see a pending volume with new change.

> Anybody who had become used to the existing behaviour and provisioned sufficient capacity only in the default AZ.

AFAIK: if AZ is not set, manila tries to use the default AZ (basically the first AZ in the database default order, e.g. if zone-2 was added before zone-1, the zone-2 will be used). If the AZ capacity is exceeded, manila won't switch to AZ2. 

Once I introduced a new SC parameter, nodbody will be affected. The backward compatibility should remain.

gouthampacha commented:
> What I'm less clear on is whether it's acceptable to change the default behaviour in a 'major' release, i.e. 1.28. In a clean installation, I'm pretty sure everybody wants the new behaviour. Ideally we would not require everybody to remember to use the featuregate flag forever to get non-broken behaviour.

another aspect of it is that this is completely opt-in. It needs the binding mode to be ``WaitForFirstConsumer`` in the storage class. Typically i've seen folks use ``Immediate`` as the binding mode with manila-csi; for example that's what is the default with storage classes created by the openshift manila csi driver operator: https://github.com/openshift/csi-driver-manila-operator/blob/8f6837f23b3d14494851cade712a01e3c242e93d/pkg/controllers/manila/manila.go#L221

jichenjc commented on docs/manila-csi-plugin/using-manila-csi-plugin.md:
```diff
@@ -50,6 +50,7 @@ Parameter | Required | Description
 `type` | _yes_ | Manila [share type](https://wiki.openstack.org/wiki/Manila/Concepts#share_type)
 `shareNetworkID` | _no_ | Manila [share network ID](https://wiki.openstack.org/wiki/Manila/Concepts#share_network)
 `availability` | _no_ | Manila availability zone of the provisioned share. If none is provided, the default Manila zone will be used. Note that this parameter is opaque to the CO and does not influence placement of workloads that will consume this share, meaning they may be scheduled onto any node of the cluster. If the specified Manila AZ is not equally accessible from all compute nodes of the c ...
+`autoTopology` | _no_ | When set to "true" and the `availability` parameter is empty, the Manila CSI controller will map the Manila availability zone to a corresponding Nova availability zone.
```
`the Manila CSI controller will map the Manila availability zone to a corresponding Nova availability zone.` 

is it to `a corresponding` or it's fixed since it's the target node (nova compute node)'s AZ ,so maybe state the info here? maybe point here to the new section you created..

kayrus replied:
I can rephrase this to a "target node".

kayrus replied:
addressed

jichenjc commented on examples/manila-csi-plugin/nfs/auto-topology-aware/storageclass.yaml:
```diff
@@ -0,0 +1,29 @@
+# We're provisioning into zone-1
+# availability parameter and allowedTopologies are empty, therefore the dynamic
+# share provisioning with automatic availability zone selection takes place.
+# The "volumeBindingMode" must be set to "WaitForFirstConsumer".
```
WaitForFirstConsumer => is this new requirement of this PR ? I think AZ related change should not change storageClass behavior like this?

kayrus replied:
I'd say it's not a requirement for this PR, but for automatic topology awareness behavior, therefore this should should reside in the storage class parameters. Do you suggest to adjust this sentence?

jichenjc replied:
no, just want to be clear on the reason of this change :)
as long as it's reasonable I am fine with it

jichenjc commented on pkg/csi/manila/controllerserver.go:
```diff
@@ -147,7 +164,8 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 	}
 
-	if err = verifyVolumeCompatibility(sizeInGiB, req, share, shareOpts, cs.d.compatOpts, shareTypeCaps); err != nil {
+	err = verifyVolumeCompatibility(sizeInGiB, req, share, shareOpts, cs.d.compatOpts, shareTypeCaps)
```
nit: seems identical, curious why change this :)

kayrus replied:
shorter lines, therefore cleaner code ;)

jichenjc commented on pkg/util/util.go:
```diff
@@ -100,3 +102,28 @@ func PatchService(ctx context.Context, client clientset.Interface, cur, mod *v1.
+	var zone string
+	var exists bool
+
+	defer func() { klog.V(4).Infof("detected AZ from the topology: %s", zone) }()
```
I think the zone log better to be V(1) so we know what's selected for trouble shooting

kayrus replied:
will do

kayrus commented on docs/manila-csi-plugin/using-manila-csi-plugin.md:
```diff
@@ -130,6 +131,32 @@ Storage AZ does not influence
 
+In cases when the Manila availability zone must correspond to the Nova
+availability zone, you can set the `autoTopology: "true"` and omit the
+`availability` parameter. By doing so, the share will be provisioned in the
```
I can address the "WaitForFirstConsumer" here. However I haven't tested the "Immediate" behavior, will try today.

kayrus commented:
I addressed the requests to this PR, let me know if there is anything else left and whether we can proceed further.

jichenjc commented:
/approve

leave lgtm to other folks to take a look especially the comments 
https://github.com/kubernetes/cloud-provider-openstack/pull/2255#issuecomment-1563083078

JoelSpeed commented:
> What I'm less clear on is whether it's acceptable to change the default behaviour in a 'major' release, i.e. 1.28. In a clean installation, I'm pretty sure everybody wants the new behaviour. Ideally we would not require everybody to remember to use the featuregate flag forever to get non-broken behaviour.

I think the change in behaviour is acceptable if you do it in a major release. You would need to make a mention of it as a bug fix in the release notes and explain the workaround for anyone who had been relying on the broken behaviour.

Proving a feature flag that allows people to preserve the old behaviour for a number of releases (normally 3?) would be preferable to prevent them being blocked on upgrades but give them a suitable timeframe (a year) to resolve the issues on their side.

Again, the feature flag and how to disable it would want to be documented in the release notes.

kayrus commented:
> I think the change in behaviour is acceptable if you do it in a major release.

Do you mean that for a major release we should set `autoTopology` to true by default? And for backward compatibility a user should set it explicitly to "false"? Do you insist on specifying a CLI argument instead of SC parameter? FYI: SC parameters cannot be updated, a SC must be recreated for that.

JoelSpeed commented:
> Do you mean that for a major release we should set autoTopology to true by default? And for backward compatibility a user should set it explicitly to "false"? 

I haven't dug into the depth of the PR to look at the mechanics, but, yes, if `autoToplogy` is your feature gate, then you would normally flip that from false to true in a 'major' release, so between 1.27 and 1.28 for example, but you wouldn't change it inline for a z-stream, eg, 1.27.2 to 1.27.3 would not be an acceptable way to change that default.

> Do you insist on specifying a CLI argument instead of SC parameter? FYI: SC parameters cannot be updated, a SC must be recreated for that.

Typically feature flags are implemented as a cluster wide thing rather than on the level of a resource, having different options for different SCs would probably be weird. I would have expect the feature flag default to be set via a CLI flag, those who need to retain old behaviour can do so across a single cluster, resolve any compatibility issues (by applying the correct details for their topology) and then flip the default right?

kayrus commented:
@JoelSpeed for me modifying the storage class is more convenient way of the feature toggling and I'd keep it in SC. Enabling the auto topology by default for the next 1.28 release sounds reasonable. I'd prefer to get at lest one more opinion before doing this.

mdbooth commented:
> @JoelSpeed for me modifying the storage class is more convenient way of the feature toggling and I'd keep it in SC. Enabling the auto topology by default for the next 1.28 release sounds reasonable. I'd prefer to get at least one more opinion before doing this.

I think this is key: what's the least disruptive way in practise to enable this behaviour change in an existing cluster? If that's changing the SC then I think the proposed design is good.

However, I thought StorageClass was immutable (but please correct me on this!). @kayrus would you be ok recreating all your storage classes to enable this feature?

mdbooth commented:
Testing, testing, 1... 2... 3..., testing. Monday, Tuesday, Wednesday.

/cherry-pick release-1.27

mdbooth commented:
> @mdbooth: once the present PR merges, I will cherry-pick it on top of release-1.27 in a new PR and assign it to you.

And I will assign it to @kayrus 😂

kayrus commented:
> However, I thought StorageClass was immutable (but please correct me on this!).

yes, it is immutable.

>  would you be ok recreating all your storage classes to enable this feature?

for old releases I'd change the default manila storage class. Looks like there is a way to do this even though there are existing PVC's.

JoelSpeed commented:
> @JoelSpeed for me modifying the storage class is more convenient way of the feature toggling and I'd keep it in SC. Enabling the auto topology by default for the next 1.28 release sounds reasonable. I'd prefer to get at least one more opinion before doing this.

I'll defer to your expertise in this area, I'm not super familiar with storage class and the relations around it

mdbooth commented:
> > However, I thought StorageClass was immutable (but please correct me on this!).
> 
> yes, it is immutable.
> 
> > would you be ok recreating all your storage classes to enable this feature?
> 
> for old releases I'd change the default manila storage class. Looks like there is a way to do this even though there are existing PVC's.
> 
> > /cherry-pick release-1.27
> 
> if we agree to set autoTopology to `true` by default for 1.28, then cherry-pick for this PR is not an option.

We can't set the default to true in this PR: it will have to be a follow-on. We should do that in any case.

mdbooth commented:
I think this is good. Adding a hold in case @kayrus wants to make additional changes from the above discussion. Please drop it if you're happy.

/lgtm
/hold for @kayrus

kayrus commented:
I'm fine with what we already have.
/unhold

kayrus commented:
/cherry-pick release-1.25
